pkg/kubelet/kuberuntime/kuberuntime_gc.go

```diff
@@ -0,0 +1,236 @@
+func (a containerByCreated) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a containerByCreated) Less(i, j int) bool { return a[i].createdAt.After(a[j].createdAt) }
+
+// toContainerGCInfo converts kubecontainer.ContainerStatus to containerGCInfo.
+func toContainerGCInfo(podUID, sandboxID string, c *kubecontainer.ContainerStatus) *containerGCInfo {
+	return &containerGCInfo{
+		containerID: c.ID.ID,
+		createdAt:   c.CreatedAt,
+		sandboxID:   sandboxID,
+		podUID:      podUID,
+	}
+}
+
+// getEvictablePods gets evictable pods.
+// Evictable pods contain sandboxes and containers which are not active and created more than MinAge ago.
+func (m *kubeGenericRuntimeManager) getEvictablePods(minAge time.Duration) (map[string]*podGCInfo, error) {
+	sandboxes, err := m.getKubeletSandboxes(true)
+	if err != nil {
+		return nil, err
+	}
+
+	evictablePods := make(map[string]*podGCInfo)
+	newestGCTime := time.Now().Add(-minAge)
+	for _, s := range sandboxes {
+		if s.GetState() != runtimeApi.PodSandBoxState_NOTREADY {
+			continue
+		}
+
+		createdAt := time.Unix(*s.CreatedAt, 0)
+		if newestGCTime.Before(createdAt) {
+			continue
+		}
+
+		podUID := getStringValueFromLabel(s.Labels, types.KubernetesPodUIDLabel)
+		if len(podUID) == 0 {
+			podUID = nonKubernetesPodUID
+		}
+		if pod, ok := evictablePods[podUID]; ok {
+			pod.sandboxIDs = append(pod.sandboxIDs, s.GetId())
+		} else {
+			evictablePods[podUID] = &podGCInfo{
+				client:     m.runtimeService,
+				containers: make([]*containerGCInfo, 0),
+				sandboxIDs: []string{s.GetId()},
+			}
+		}
+
+		containerStatuses, err := m.getKubeletContainerStatuses(s.GetId())
+		if err != nil {
+			return nil, err
+		}
+		for _, c := range containerStatuses {
+			evictablePods[podUID].containers = append(evictablePods[podUID].containers,
+				toContainerGCInfo(podUID, s.GetId(), c))
+		}
+
+		sort.Sort(containerByCreated(evictablePods[podUID].containers))
+	}
+
+	return evictablePods, nil
+}
+
+// enforceMaxPerPodContainer enforces max number of dead containers of each pod and
+// returns the number of total evictable containers.
+func enforceMaxPerPodContainer(evictablePods map[string]*podGCInfo, maxPerPodContainer int) int {
+	totalEvictableContainers := 0
+	for _, pod := range evictablePods {
+		if maxPerPodContainer > 0 {
+			pod.removeOldestNContainers(pod.containersNum() - maxPerPodContainer)
+		}
+		totalEvictableContainers += pod.containersNum()
+	}
+
+	return totalEvictableContainers
+}
+
+// GarbageCollect removes dead containers using the specified container gc policy.
+// It consists of following steps:
+// * gets evictable pods with belonging containers and sandboxes which are not active and created more than gcPolicy.MinAge ago.
+// * removes oldest dead containers for each pod by enforcing gcPolicy.MaxPerPodContainer.
+// * removes oldest dead containers by enforcing gcPolicy.MaxContainers.
+// * removes dead sandboxes with zero containers.
```

Comment:
Please make it clear that we don't count sandboxes when enforcing the GC policy.